File: platform/app/.recipes/Nginx-Orthanc-Postgres/config/on-stored-instance.lua
function OnStoredInstance(instanceId, tags, metadata)
  -- Extract DICOM tags from the instance
  local instance = RestApiGet('/instances/' .. instanceId)
  local parsed = ParseJson(instance)
  local mainTags = parsed['MainDicomTags'] or {}

  -- AccessionNumber is a study-level tag; fetch from parent study
  local parentStudyId = parsed['ParentStudy']
  local accessionNumber = nil
  if parentStudyId then
    local study = ParseJson(RestApiGet('/studies/' .. parentStudyId))
    local studyTags = study['MainDicomTags'] or {}
    accessionNumber = studyTags['AccessionNumber']
  end

  -- Build payload for the PACS server
  local payload = {
    orthancId = instanceId,
    sopInstanceUid = mainTags['SOPInstanceUID'],
    seriesInstanceUid = mainTags['SeriesInstanceUID'] or parsed['ParentSeries'],
    studyInstanceUid = mainTags['StudyInstanceUID'] or parentStudyId,
    sopClassUid = mainTags['SOPClassUID'],
    modality = mainTags['Modality'],
    patientId = mainTags['PatientID'],
    patientName = mainTags['PatientName'],
    accessionNumber = accessionNumber,
  }

  -- POST to PACS NestJS server
  local instancesUrl = os.getenv('ORTHANC_INSTANCES_URL') or 'http://host.docker.internal:3002/api/orthanc/instances'
  local apiKey = os.getenv('ORTHANC_API_KEY') or 'dev-orthanc-key'
  local headers = {
    ['Content-Type'] = 'application/json',
    ['x-api-key'] = apiKey,
  }

  -- Use pcall to catch errors without blocking Orthanc
  local success, err = pcall(function()
    HttpPost(instancesUrl,
             DumpJson(payload),
             headers)
  end)

  if not success then
    PrintToStdout('OnStoredInstance: Failed to notify PACS server for instance ' .. instanceId .. ': ' .. tostring(err))
  else
    PrintToStdout('OnStoredInstance: Notified PACS server for instance ' .. instanceId)
  end
end

File: platform/app/.recipes/Nginx-Orthanc-Postgres/config/on-stored-instance.py
import orthanc
import json
import os
import asyncio
from bullmq import Queue

REDIS_HOST = os.environ.get('REDIS_HOST', 'redis')
REDIS_PORT = int(os.environ.get('REDIS_PORT', '6379'))
QUEUE_NAME = 'dicom-instances'
QUEUE_OPTS = {
    'connection': {
        'host': REDIS_HOST,
        'port': REDIS_PORT,
    }
}


async def enqueue(payload):
    queue = Queue(QUEUE_NAME, QUEUE_OPTS)
    await queue.add('instance-stored', payload, {
        'attempts': 5,
        'backoff': {'type': 'exponential', 'delay': 1000},
    })
    await queue.close()


def OnStoredInstance(dicom, instanceId):
    try:
        tags = json.loads(dicom.GetInstanceSimplifiedJson())
        payload = {
            'orthancId': instanceId,
            'sopInstanceUid': tags.get('SOPInstanceUID'),
            'seriesInstanceUid': tags.get('SeriesInstanceUID'),
            'studyInstanceUid': tags.get('StudyInstanceUID'),
            'sopClassUid': tags.get('SOPClassUID'),
            'modality': tags.get('Modality'),
            'patientId': tags.get('PatientID'),
            'patientName': tags.get('PatientName'),
        }
        asyncio.run(enqueue(payload))
        orthanc.LogInfo(f'Queued instance {instanceId} to Redis')
    except Exception as e:
        orthanc.LogError(f'Failed to queue instance {instanceId}: {e}')


orthanc.RegisterOnStoredInstanceCallback(OnStoredInstance)
